File: src/services/timeline.ts
/**
 * API client methods for timeline-related endpoints.
 *
 * Provides typed functions for fetching timeline data and trend analysis
 * for temporal visualization of case statistics.
 */

import { apiClient } from './api'
import { FilterState } from '../types/filter'
import { buildQueryParams } from './cases'
import {
  TimelineDataResponse,
  TimelineTrendResponse,
  TimelineGranularity,
  TimelineMetric,
  TimelineDataParams,
  TimelineTrendParams,
} from '../types/timeline'

/**
 * Build timeline-specific query parameters from filter state.
 *
 * Extends the base query params with timeline-specific options.
 *
 * @param filters - Filter state from Zustand store
 * @param granularity - Time granularity (year, month, decade)
 * @returns API-compatible query parameters for timeline endpoints
 */
const buildTimelineParams = (
  filters: FilterState,
  granularity: TimelineGranularity = 'year'
): TimelineDataParams => {
  const baseParams = buildQueryParams(filters)

  return {
    granularity,
    year_start: filters.yearRange[0],
    year_end: filters.yearRange[1],
    state_code: filters.states.length === 1 ? filters.states[0] : undefined,
    solved_status: filters.solved !== 'all' ? filters.solved : undefined,
  }
}

/**
 * Fetch timeline data for case statistics over time.
 *
 * Returns aggregated case counts (total, solved, unsolved) grouped by
 * the specified time granularity.
 *
 * @param filters - Filter criteria from the filter store
 * @param granularity - Time granularity: 'year', 'month', or 'decade'
 * @returns Promise resolving to timeline data response
 */
export const fetchTimelineData = async (
  filters: FilterState,
  granularity: TimelineGranularity = 'year'
): Promise<TimelineDataResponse> => {
  const params = buildTimelineParams(filters, granularity)

  const response = await apiClient.get<TimelineDataResponse>('/api/timeline/data', {
    params,
  })
  return response.data
}

/**
 * Fetch trend analysis data for a specific metric.
 *
 * Returns time series data with optional moving average calculation
 * for trend visualization.
 *
 * @param filters - Filter criteria from the filter store
 * @param metric - Metric to analyze: 'solve_rate', 'total_cases', etc.
 * @param granularity - Time granularity: 'year', 'month', or 'decade'
 * @param movingAverageWindow - Window size for moving average (default: 3)
 * @returns Promise resolving to trend response
 */
export const fetchTimelineTrends = async (
  filters: FilterState,
  metric: TimelineMetric = 'solve_rate',
  granularity: TimelineGranularity = 'year',
  movingAverageWindow: number = 3
): Promise<TimelineTrendResponse> => {
  const baseParams = buildTimelineParams(filters, granularity)

  const params: TimelineTrendParams = {
    ...baseParams,
    metric,
    moving_average_window: movingAverageWindow,
  }

  const response = await apiClient.get<TimelineTrendResponse>('/api/timeline/trends', {
    params,
  })
  return response.data
}

/**
 * Fetch timeline data for a specific state.
 *
 * Convenience method for state-level timeline analysis.
 *
 * @param stateCode - State code (e.g., "CA", "TX")
 * @param granularity - Time granularity
 * @param yearStart - Start year for the range
 * @param yearEnd - End year for the range
 * @returns Promise resolving to timeline data response
 */
export const fetchStateTimeline = async (
  stateCode: string,
  granularity: TimelineGranularity = 'year',
  yearStart: number = 1976,
  yearEnd: number = 2023
): Promise<TimelineDataResponse> => {
  const params: TimelineDataParams = {
    granularity,
    year_start: yearStart,
    year_end: yearEnd,
    state_code: stateCode,
  }

  const response = await apiClient.get<TimelineDataResponse>('/api/timeline/data', {
    params,
  })
  return response.data
}

/**
 * Fetch comparison timeline data for multiple states.
 *
 * Useful for comparing trends across different regions.
 *
 * @param stateCodes - Array of state codes to compare
 * @param granularity - Time granularity
 * @param yearStart - Start year for the range
 * @param yearEnd - End year for the range
 * @returns Promise resolving to array of timeline responses
 */
export const fetchStateComparison = async (
  stateCodes: string[],
  granularity: TimelineGranularity = 'year',
  yearStart: number = 1976,
  yearEnd: number = 2023
): Promise<Map<string, TimelineDataResponse>> => {
  const results = new Map<string, TimelineDataResponse>()

  // Fetch data for each state in parallel
  const promises = stateCodes.map(async (stateCode) => {
    const data = await fetchStateTimeline(stateCode, granularity, yearStart, yearEnd)
    return { stateCode, data }
  })

  const responses = await Promise.all(promises)

  responses.forEach(({ stateCode, data }) => {
    results.set(stateCode, data)
  })

  return results
}

File: backend/routes/timeline.py
"""Timeline API routes for temporal visualization.

Provides endpoints for timeline data aggregation and trend analysis
with support for various time granularities and filters.
"""

import logging
from typing import Literal, Optional

from fastapi import APIRouter, Query

from models.timeline import TimelineDataResponse, TimelineTrendResponse
from services.timeline_service import get_timeline_data, get_timeline_trends

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/timeline", tags=["timeline"])


@router.get("/data", response_model=TimelineDataResponse)
async def timeline_data(
    granularity: Literal["year", "month", "decade"] = Query(
        default="year",
        description="Time aggregation granularity"
    ),
    state: Optional[str] = Query(
        default=None,
        description="Filter by state name"
    ),
    county: Optional[str] = Query(
        default=None,
        description="Filter by county FIPS code"
    ),
    year_start: Optional[int] = Query(
        default=None,
        ge=1965,
        le=2030,
        description="Start year (inclusive)"
    ),
    year_end: Optional[int] = Query(
        default=None,
        ge=1965,
        le=2030,
        description="End year (inclusive)"
    ),
    solved: Optional[bool] = Query(
        default=None,
        description="Filter by solved status"
    ),
    victim_sex: Optional[str] = Query(
        default=None,
        description="Filter by victim sex"
    ),
    victim_race: Optional[str] = Query(
        default=None,
        description="Filter by victim race"
    ),
    victim_age_min: Optional[int] = Query(
        default=None,
        ge=0,
        le=999,
        description="Minimum victim age"
    ),
    victim_age_max: Optional[int] = Query(
        default=None,
        ge=0,
        le=999,
        description="Maximum victim age"
    ),
    weapon: Optional[str] = Query(
        default=None,
        description="Filter by weapon type"
    ),
    relationship: Optional[str] = Query(
        default=None,
        description="Filter by relationship"
    ),
    circumstance: Optional[str] = Query(
        default=None,
        description="Filter by circumstance"
    ),
) -> TimelineDataResponse:
    """Get aggregated case data by time period for timeline visualization.
    
    Aggregates cases by the specified time granularity (year, month, or decade),
    calculating total cases, solved/unsolved counts, and solve rates.
    
    Args:
        granularity: Time aggregation level ("year", "month", or "decade")
        state: Filter by state name
        county: Filter by county FIPS code
        year_start: Filter by start year (inclusive)
        year_end: Filter by end year (inclusive)
        solved: Filter by solved status
        victim_sex: Filter by victim sex
        victim_race: Filter by victim race
        victim_age_min: Filter by minimum victim age
        victim_age_max: Filter by maximum victim age
        weapon: Filter by weapon type
        relationship: Filter by relationship
        circumstance: Filter by circumstance
        
    Returns:
        TimelineDataResponse with aggregated data points and metadata
        
    Example:
        GET /api/timeline/data?granularity=year&state=California&year_start=2000
    """
    logger.info(
        f"Timeline data request: granularity={granularity}, "
        f"state={state}, year_start={year_start}, year_end={year_end}"
    )
    
    return get_timeline_data(
        granularity=granularity,
        state=state,
        county=county,
        year_start=year_start,
        year_end=year_end,
        solved=solved,
        victim_sex=victim_sex,
        victim_race=victim_race,
        victim_age_min=victim_age_min,
        victim_age_max=victim_age_max,
        weapon=weapon,
        relationship=relationship,
        circumstance=circumstance,
    )


@router.get("/trends", response_model=TimelineTrendResponse)
async def timeline_trends(
    metric: Literal["solve_rate", "total_cases", "unsolved_cases", "solved_cases"] = Query(
        default="solve_rate",
        description="Metric to analyze"
    ),
    granularity: Literal["year", "month", "decade"] = Query(
        default="year",
        description="Time aggregation granularity"
    ),
    moving_average_window: int = Query(
        default=3,
        ge=2,
        le=10,
        description="Window size for moving average calculation"
    ),
    state: Optional[str] = Query(
        default=None,
        description="Filter by state name"
    ),
    county: Optional[str] = Query(
        default=None,
        description="Filter by county FIPS code"
    ),
    year_start: Optional[int] = Query(
        default=None,
        ge=1965,
        le=2030,
        description="Start year (inclusive)"
    ),
    year_end: Optional[int] = Query(
        default=None,
        ge=1965,
        le=2030,
        description="End year (inclusive)"
    ),
    solved: Optional[bool] = Query(
        default=None,
        description="Filter by solved status"
    ),
    victim_sex: Optional[str] = Query(
        default=None,
        description="Filter by victim sex"
    ),
    victim_race: Optional[str] = Query(
        default=None,
        description="Filter by victim race"
    ),
    victim_age_min: Optional[int] = Query(
        default=None,
        ge=0,
        le=999,
        description="Minimum victim age"
    ),
    victim_age_max: Optional[int] = Query(
        default=None,
        ge=0,
        le=999,
        description="Maximum victim age"
    ),
    weapon: Optional[str] = Query(
        default=None,
        description="Filter by weapon type"
    ),
    relationship: Optional[str] = Query(
        default=None,
        description="Filter by relationship"
    ),
    circumstance: Optional[str] = Query(
        default=None,
        description="Filter by circumstance"
    ),
) -> TimelineTrendResponse:
    """Get trend analysis data with moving averages.
    
    Calculates trends for the specified metric over time, including
    moving averages for smoothing. Useful for identifying patterns
    and long-term trends in the data.
    
    Args:
        metric: Metric to analyze ("solve_rate", "total_cases", "unsolved_cases", "solved_cases")
        granularity: Time aggregation level ("year", "month", or "decade")
        moving_average_window: Window size for moving average (2-10)
        state: Filter by state name
        county: Filter by county FIPS code
        year_start: Filter by start year (inclusive)
        year_end: Filter by end year (inclusive)
        solved: Filter by solved status
        victim_sex: Filter by victim sex
        victim_race: Filter by victim race
        victim_age_min: Filter by minimum victim age
        victim_age_max: Filter by maximum victim age
        weapon: Filter by weapon type
        relationship: Filter by relationship
        circumstance: Filter by circumstance
        
    Returns:
        TimelineTrendResponse with trend data points and moving averages
        
    Example:
        GET /api/timeline/trends?metric=solve_rate&granularity=year&moving_average_window=5
    """
    logger.info(
        f"Timeline trends request: metric={metric}, granularity={granularity}, "
        f"window={moving_average_window}, state={state}"
    )
    
    return get_timeline_trends(
        metric=metric,
        granularity=granularity,
        moving_average_window=moving_average_window,
        state=state,
        county=county,
        year_start=year_start,
        year_end=year_end,
        solved=solved,
        victim_sex=victim_sex,
        victim_race=victim_race,
        victim_age_min=victim_age_min,
        victim_age_max=victim_age_max,
        weapon=weapon,
        relationship=relationship,
        circumstance=circumstance,
    )